### ISSUE TITLE
Failed LDS update results in invalid configuration state

### ISSUE THREAD

**Issue by murray-stripe** [envoyproxy/envoy#12592]
*Title*: __Failed LDS update results in invalid configuration state__

_**Description**_:
Our production control-plane experienced and issue that caused us to deliver an invalid LDS update. Specifically this update deleted one listener and added a new listener that contained a configuration that passed protobuf validation but failed when [creating the listener][creation_exception]. The xDS API returned and error and reported the current version as the previous good version. However, at this point, the listener that was removed by the invalid update is gone.

We detected this error in our control-plane, removed the invalid update, and restart our config server. It delivered an LDS update on start which was the same version as Envoy currently advertises as having (despite the missing listener that was removed by the invalid update). Envoy doesn't process the update, likely because it appears to be a duplicate.

It took us a while to eventually find what was prolonging the issue as both now Envoy and our control plane advertised the same configuration version. Eventually Envoy was restarted which properly consumed the LDS update and the issue was resolved.

_**Repro steps**_:

This was reproduced with our in-house control-plane, so I do not have code to share, but generally the following updates should reproduce the issue.

0. Start Envoy
1. Deliver first LDS update

```json
{ "name": "listener-1",
  "address": {
    "socket_address": { "address": "0.0.0.0", "port_value": "8080" }
  }
}
```

2. Validate everything is functional and listener is setup
3. Deliver second LDS update (remove `listener-1` + add `listener-2` with an invalid address)

```json
{ "name": "listener-2",
  "address": {
    "socket_address": { "address": "300.1.1.1", "port_value": "8080" }
  }
}
```

4. This configuration will pass protobuf validation but fail to be applied. `listener-1` is no longer active.

At this point you can attempt to re-apply the original LDS update, `listener-1`, but Envoy will see these updates as identical to the current version. The only way to recover from this invalid configuration is to restart the Envoy process.

_**Additional Notes/Thoughts**_

I dug through the code starting [here][lds_api] and identified the following cases in which an exception is thrown:

 + Pipe address is longer than 108 characters
 + A mode is specified on the pipe
 + Pipe path with null-terminators included in it
 + An invalid resolver name is provided (Socket Address)
 + A named port is provided with the default resolver
 + An invalid IPv4 or IPv6 address is given
 + Setting socket options fail

Aside from the last one, I believe most of these exceptions could be pushed into a "higher layer" and performed before Envoy starts applying configuration changes. This doesn't solve the problem per se, but it does address a lot of errors that would land someone in this situation.

  [creation_exception]: https://github.com/envoyproxy/envoy/blob/master/source/server/lds_api.cc#L76-L82
  [lds_api]: https://github.com/envoyproxy/envoy/blob/bb3571a6207820ad0de4ef9be6db838e175f6054/source/server/lds_api.cc#L70

**Comment by mattklein123** [comment]
cc @lambdai @htuch 

I think this is the general problem that we have where listeners can fail when they are applied to a worker, but after XDS ACK. There may be other bugs here also. @ggreenway had the idea to perform more processing on the main thread before we ack which I think would make the number of possible errors that can happen on the worker much smaller. It's probably worth doing this.

**Comment by htuch** [comment]
I think the general solution is https://github.com/envoyproxy/envoy/issues/11396 and allowing the server to subscribe to the state of listeners on the client. Today, the best you can do is query stats, which is OOB for xDS.

**Comment by lambdai** [comment]
I see another problem: the execution of the new configuration is not atomic.
When the config A is updating the config B 
Regardless what the config version (A or B) is used, the reality is that the execution stopped at somewhere between A and B.
While transaction support commit and rollback, which we probably wont' support b/c best efforts are also preferred in certain cases

**Comment by lambdai** [comment]
In some way, the state between config A and config B
listener manager may want to see config unknown so that the new update should always be allowed regardless the ack/nack was sent to xDS server.

So that at least the new config is never rejected, at the cost of duplicated update. The good new is that duplicated update is somewhat addressed by smart listener update.

**Comment by htuch** [comment]
Wondering what is actionable here. https://github.com/envoyproxy/envoy/issues/12592#issuecomment-673031897 has some suggestions, but I'm not sure this is something that will save us in all cases, #11396 is the long term fix.

**Comment by lambdai** [comment]
This should be fixed by #12645

**Comment by lambdai** [comment]
Well, not a complete solution, but manually triggered rollback is possible
